### ISSUE TITLE
Front End modularization How to achieve？

### ISSUE THREAD

**Issue by fendo8888** [hank-cp/sbp#1]
Front End modularization How to achieve？

**Comment by hank-cp** [comment]
1. for old school server renderred via template engine, like jsp/structs/FreeMarker/, etc, since sbp has handled loading static resource inside of plugin, this part should be ready. I am just running out of time to provide examples. Will do it in some days later.

2. for morden front/server separation project, I am still looking for mature solution of dynamically pluggabla framework. For our own production project using sbp right now, it's still built into a whole bundle. I know there is technology dynamically load pluggable bundle via iframe, but there is a lot down side of this approach.

Please let me know if you have any suggestion~

**Comment by fendo8888** [comment]
Fucking great！If the template engine is used, how to extend the page like hook in WordPress?

**Comment by hank-cp** [comment]
To make you app extensible, an extension registration framework is needed. You could take a look at my another project [skr](https://github.com/hank-cp/skr), wihch provides skeleton of auth/security/extension registration. This project is in kinds of incubation state right now, but it also has been used in some of our production projects.

**Comment by fendo8888** [comment]
好的

**Comment by syifarahmat** [comment]
I use extjs but extjs is commercial 😁😁
https://stackoverflow.com/questions/29763532/dynamically-load-extjs-modular-application

**Comment by hank-cp** [comment]
[Webpack 5 module federation](https://indepth.dev/webpack-5-module-federation-a-game-changer-in-javascript-architecture/) might be the key. Will take a close look at it.

**Comment by fendo8888** [comment]
@hank-cp Can Webpack 5 module federation combine pf4j to achieve modular development and extension points?

**Comment by hank-cp** [comment]
> @hank-cp Can Webpack 5 module federation combine pf4j to achieve modular development and extension points?

No. module federation is totally web technology that nothing relates to pf4j, which is totally server side technoalogy. If it helps, it will be another **sbp-vue-client** or **sbp-angular-client** or **sbp-reactjs-client** project to manage front end modularization specifically.

**Comment by ScriptedAlchemy** [comment]
Actually, Module Federation works with anything that uses the JavaScript environment runtime. Right now I’ve federated middleware across servers, my GraphQL layer uses module federation against 6 independently deployed systems. 

Not sure of the context of this project, but as the creator of Module Federation, I know what will do. If you’ve got questions, I’m happy to discuss. 

What is the need here and use case for this project?

**Comment by j-barata** [comment]
Maybe this kind of library may also be worth taking a look: [remote-component](https://github.com/Paciolan/remote-component) or [remote-module-loader](https://github.com/Paciolan/remote-module-loader) ?

**Comment by ScriptedAlchemy** [comment]
That solution still is more complex than module federation. Which is built into webpack itself. It's designed to let you import from other bundles deployed on different URLs, client or server - module federation supports this. You simply import remote components like you would anything else. 

https://github.com/module-federation/module-federation-examples/tree/master/server-side-render-only

Module federation also works on projects like next js, though the server implementation in v9 requires a workaround. V10 will support it out the box. SSR and all. 

Just sayin, what you want is built-in to webpack and works as expected in production for the past 6 months 🤷‍♂️

**Comment by j-barata** [comment]
Hi,
Module Federation is a very interesting technology indeed.
As far as I understood, if an application A wants to use two remote modules B and C, they both need to be declared in the webpack config of the application A, right ?
What happens if another plugin (module) D is installed on the application A at runtime (i.e without updating the webpack config and rebuilding it) ?

But maybe I missed something ?

**Comment by ScriptedAlchemy** [comment]
Dynamic remotes are supported and I've used them in several ways. In the module federation examples repo, there's a dynamic remote sample doing just that. Remotes can also be nested and circular without problem

**Comment by j-barata** [comment]
You're probably referring to this feature : [dynamic-remote-containers](https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers) ?
I hadn't seen that feature yet, thank you for mentioning it :) looks great
